R/parametric.R

```diff
@@ -0,0 +1,363 @@
+#' @param distr6 (`logical(1)`)\cr
+#' If `FALSE` (default) and `type` is `"survival"` or `"all"` returns matrix of survival
+#' probabilities, otherwise returns a [distr6::Matdist()].
+#' @param ntime `(numeric(1))`\cr
+#' Number of unique time-points in the training set, default is 150.
+#' @param round_time `(numeric(1))`\cr
+#' Number of decimal places to round time-points to, default is 2, set to `FALSE` for no rounding.
+#' @param ... `ANY` \cr
+#' Currently ignored.
+#'
+#'
+#' @examples
+#' if (requireNamespaces(c("distr6", "survival"))) {
+#'   library(survival)
+#'
+#'   set.seed(42)
+#'   train <- simsurvdata(50)
+#'   test <- simsurvdata(50)
+#'   fit <- parametric(Surv(time, status) ~ ., data = train)
+#'   predict(fit, newdata = test)
+#'
+#'   # distr6 parameter ignored when returning continuous distribution
+#'   predict_distr <- predict(fit, newdata = test, distr6 = FALSE)
+#'   predict_distr$survival(3)
+#'
+#'   # Return a discrete distribution survival matrix
+#'   predict_distr <- predict(fit, newdata = test,
+#'     return_method = "discrete", distr6 = FALSE)
+#'   predict_distr[1:5, 1:5]
+#'
+#'   # Return a relative risk ranking with type = "risk"
+#'   predict(fit, newdata = test, type = "risk")[1:5]
+#'
+#'   # Or survival probabilities and a rank
+#'   predict(fit, newdata = test, type = "all", distr6 = TRUE)
+#' }
+#' @export
+predict.parametric <- function(object, newdata,
+  form = c("aft", "ph", "tobit", "po"), times = NULL,
+  return_method = c("continuous", "discrete"),
+  type = c("survival", "risk", "all"), distr6 = FALSE,
+  ntime = 150, round_time = 2, ...) {
+
+  return_method <- match.arg(return_method)
+  form <- match.arg(form)
+  type <- match.arg(type)
+
+  unique_times <- sort(unique(object$model$y[, 1, drop = FALSE]))
+  if (!is.logical(round_time) || round_time) {
+    unique_times <- unique(round(unique_times, round_time))
+  }
+  # using same method as in ranger
+  unique_times <- unique_times[
+    unique(round(seq.int(1, length(unique_times), length.out = ntime)))
+  ]
+
+  truth <- object$model$y
+  newdata <- clean_test_data(object$model, newdata)
+
+  if (is.null(times)) {
+    predict_times <- unique_times
+  } else {
+    predict_times <- sort(unique(times))
+  }
+
+  basedist <- object$basedist
+  fit <- object$model
+  lp <- matrix(fit$coefficients[-1], nrow = 1) %*% t(newdata)
+
+  if (return_method == "discrete") {
+    surv <- .predict_survreg_discrete(object, newdata, form, predict_times,
+      basedist, fit, lp)
+  } else {
+    surv <- .predict_survreg_continuous(object, newdata, form,
+      basedist, fit, lp)
+  }
+
+  ret <- list()
+
+  if (type %in% c("risk", "all")) {
+    ret$risk <- -surv$lp
+  }
+
+  if (type %in% c("survival", "all")) {
+    if (distr6 || return_method == "continuous") {
+      ret$surv <- surv$distr
+    } else {
+      ret$surv <- 1 - distr6::gprm(surv$distr, "cdf")
```

Comment:
I would simplify and optimize here: first, the parameter `distr6` should not do anything for this learner so that the `return_method` can clearly define the output. Second, here we re-convert the supposedly discrete result (which is now a `distr6`) back to a survival matrix, which could easily avoid doing